File: src/main/java/au/id/micolous/metrodroid/fragment/CardBalanceFragment.java
/*
 * CardBalanceFragment.java
 *
 * Copyright 2012-2014 Eric Butler <eric@codebutler.com>
 * Copyright 2015 Michael Farrell <micolous+git@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.fragment;

import android.app.ListFragment;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.content.res.AppCompatResources;
import android.text.Spanned;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;

import au.id.micolous.metrodroid.activity.CardInfoActivity;
import au.id.micolous.metrodroid.transit.Subscription;
import au.id.micolous.metrodroid.transit.TransitBalance;
import au.id.micolous.metrodroid.transit.TransitData;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import au.id.micolous.farebot.R;
import au.id.micolous.metrodroid.transit.TransitCurrency;
import au.id.micolous.metrodroid.ui.ListItem;
import au.id.micolous.metrodroid.util.TripObfuscator;
import au.id.micolous.metrodroid.util.Utils;

public class CardBalanceFragment extends ListFragment {
    private TransitData mTransitData;
    private static final String TAG = "CardBalanceFragment";

    private static final String TAG_BALANCE_VIEW = "balanceView";
    private static final String TAG_SUBSCRIPTION_VIEW = "subscriptionView";
    private static final String TAG_ERROR_VIEW = "errorView";

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mTransitData = getArguments().getParcelable(CardInfoActivity.EXTRA_TRANSIT_DATA);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ArrayList<Object> combined = new ArrayList<>();
        List<TransitBalance> balances = mTransitData.getBalances();
        if (balances != null)
            combined.addAll(balances);
        List<? extends Subscription> subscriptions = mTransitData.getSubscriptions();
        if (subscriptions != null)
            combined.addAll(subscriptions);
        setListAdapter(new BalancesAdapter(getActivity(), combined));
    }

    private class BalancesAdapter extends ArrayAdapter<Object> {
        public BalancesAdapter(Context context, List<Object> balances) {
            super(context, 0, balances);
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            Object item = getItem(position);

            if (item == null) {
                // https://github.com/micolous/metrodroid/issues/28
                Log.w(TAG, "null balance received -- this is an error");
                return getErrorView(convertView, parent, "null");
            }

            if (item instanceof TransitBalance)
                return getBalanceView(convertView, parent, (TransitBalance) item);

            if (item instanceof Subscription)
                return getSubscriptionView(convertView, parent, (Subscription) item);
            return getErrorView(convertView, parent, item.getClass().getSimpleName());
        }

        private View getErrorView(View convertView, ViewGroup parent, String err) {
            View view = convertView;
            if (view == null || view.getTag() != TAG_ERROR_VIEW) {
                view = getActivity().getLayoutInflater().inflate(R.layout.balance_item, parent, false);
                view.setTag(TAG_ERROR_VIEW);
            }

            ((TextView) view.findViewById(R.id.balance)).setText(err);
            return view;
        }

        public View getSubscriptionView(View convertView, ViewGroup parent, Subscription subscription) {
            View view = convertView;
            if (view == null || view.getTag() != TAG_SUBSCRIPTION_VIEW) {
                view = getActivity().getLayoutInflater().inflate(R.layout.subscription_item, parent, false);
                view.setTag(TAG_SUBSCRIPTION_VIEW);
            }

            TextView validView = view.findViewById(R.id.valid);
            if (subscription.getValidFrom() != null && subscription.getValidTo() != null) {
                Spanned validFrom = Utils.dateFormat(TripObfuscator.maybeObfuscateTS(subscription.getValidFrom()));
                Spanned validTo;
                if (subscription.validToHasTime()) {
                    validTo = Utils.dateTimeFormat(TripObfuscator.maybeObfuscateTS(subscription.getValidTo()));
                } else {
                    validTo = Utils.dateFormat(TripObfuscator.maybeObfuscateTS(subscription.getValidTo()));
                }
                validView.setText(getString(R.string.valid_format, validFrom, validTo));
                validView.setVisibility(View.VISIBLE);
            } else if (subscription.getValidTo() != null) {
                Spanned validTo;
                if (subscription.validToHasTime()) {
                    validTo = Utils.dateTimeFormat(TripObfuscator.maybeObfuscateTS(subscription.getValidTo()));
                } else {
                    validTo = Utils.dateFormat(TripObfuscator.maybeObfuscateTS(subscription.getValidTo()));
                }
                validView.setText(getString(R.string.valid_to_format, validTo));
                validView.setVisibility(View.VISIBLE);
            } else if (subscription.getValidFrom() != null) {
                Spanned validTo = Utils.dateFormat(TripObfuscator.maybeObfuscateTS(subscription.getValidFrom()));
                validView.setText(getString(R.string.valid_from_format, validTo));
                validView.setVisibility(View.VISIBLE);
            } else {
                validView.setVisibility(View.GONE);
            }

            TextView tripsView = view.findViewById(R.id.trips);
            Integer remainingTrips = subscription.getRemainingTripCount();
            Integer totalTrips = subscription.getTotalTripCount();

            if (remainingTrips != null && totalTrips != null) {
                tripsView.setText(Utils.localizePlural(R.plurals.trips_remaining_total,
                        remainingTrips, remainingTrips, totalTrips));
                tripsView.setVisibility(View.VISIBLE);
            } else if (remainingTrips != null) {
                tripsView.setText(Utils.localizePlural(R.plurals.trips_remaining,
                        remainingTrips, remainingTrips));
                tripsView.setVisibility(View.VISIBLE);
            } else {
                tripsView.setVisibility(View.GONE);
            }

            TextView companyView = view.findViewById(R.id.company);
            String agencyName = subscription.getAgencyName(true);
            if (agencyName != null) {
                companyView.setText(agencyName);
                companyView.setVisibility(View.VISIBLE);
            } else
                companyView.setVisibility(View.GONE);
            TextView nameView = view.findViewById(R.id.name);
            String name = subscription.getSubscriptionName();
            if (name != null)  {
                nameView.setText(name);
                nameView.setVisibility(View.VISIBLE);
            } else {
                nameView.setVisibility(View.GONE);
            }

            // TODO: Replace this with structured data.
            TextView usedView = view.findViewById(R.id.used);
            if (subscription.getSubscriptionState() == Subscription.SubscriptionState.UNKNOWN) {
                usedView.setVisibility(View.GONE);
            } else {
                usedView.setText(subscription.getSubscriptionState().getDescription());
                usedView.setVisibility(View.VISIBLE);
            }

            LinearLayout paxLayout = view.findViewById(R.id.pax_layout);
            ImageView paxIcon = view.findViewById(R.id.pax_icon);
            TextView paxTextView = view.findViewById(R.id.pax_text_view);
            int pax = subscription.getPassengerCount();

            if (pax >= 1) {
                paxTextView.setText(String.format(Locale.getDefault(), "%d", pax));
                paxIcon.setContentDescription(Utils.localizePlural(R.plurals.passengers, pax));

                paxIcon.setImageDrawable(AppCompatResources.getDrawable(getContext(),
                        pax == 1 ? R.drawable.material_ic_person_24dp : R.drawable.material_ic_group_24dp));

                paxLayout.setVisibility(View.VISIBLE);
            } else {
                // No information.
                paxLayout.setVisibility(View.GONE);
            }

            boolean hasExtraInfo = subscription.getInfo() != null;
            ListView properties = view.findViewById(R.id.properties);
            TextView moreInfoPrompt = view.findViewById(R.id.more_info_prompt);

            if (hasExtraInfo) {
                moreInfoPrompt.setVisibility(View.VISIBLE);
                properties.setVisibility(View.GONE);
            } else {
                properties.setVisibility(View.GONE);
                moreInfoPrompt.setVisibility(View.GONE);
            }

            return view;
        }

        private View getBalanceView(View convertView,
                                    ViewGroup parent, TransitBalance balance) {
            View view = convertView;
            if (view == null || view.getTag() != TAG_BALANCE_VIEW) {
                view = getActivity().getLayoutInflater().inflate(R.layout.balance_item, parent, false);
                view.setTag(TAG_BALANCE_VIEW);
            }

            TextView validView = view.findViewById(R.id.valid);
            if (balance.getValidFrom() != null && balance.getValidTo() != null) {
                Spanned validFrom = Utils.dateFormat(TripObfuscator.maybeObfuscateTS(balance.getValidFrom()));
                Spanned validTo = Utils.dateFormat(TripObfuscator.maybeObfuscateTS(balance.getValidTo()));
                validView.setText(getString(R.string.valid_format, validFrom, validTo));
                validView.setVisibility(View.VISIBLE);
            } else if (balance.getValidTo() != null) {
                Spanned validTo = Utils.dateFormat(TripObfuscator.maybeObfuscateTS(balance.getValidTo()));
                validView.setText(getString(R.string.valid_to_format, validTo));
                validView.setVisibility(View.VISIBLE);
            } else {
                validView.setVisibility(View.GONE);
            }

            String name = balance.getName();
            TextView nameView = view.findViewById(R.id.name);
            TextView balanceView = view.findViewById(R.id.balance);
            TransitCurrency balanceCur = balance.getBalance();
            if (name != null) {
                nameView.setText(name);
                nameView.setVisibility(View.VISIBLE);
            } else
                nameView.setVisibility(View.GONE);
            if (balanceCur != null) {
                balanceView.setText(balanceCur.maybeObfuscateBalance().formatCurrencyString(true));
                balanceView.setVisibility(View.VISIBLE);
            } else
                balanceView.setVisibility(View.GONE);

            return view;
        }

        @Override
        public boolean isEnabled(int position) {
            Object item = getItem(position);

            if (item == null) {
                return false;
            }

            if (item instanceof TransitBalance) {
                // We don't do anything for balances, yet.
                return false;
            }

            if (item instanceof Subscription) {
                return ((Subscription) item).getInfo() != null;
            }

            return false;
        }
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id) {
        Log.d(TAG, "Clicked " + id + " " + position );
        Object item = getListAdapter().getItem(position);
        if (item == null) {
            return;
        }

        if (item instanceof TransitBalance) {
            return;
        }

        if (item instanceof Subscription) {
            List<ListItem> infos = ((Subscription)item).getInfo();
            if (infos == null) {
                return;
            }

            ListView lv = v.findViewById(R.id.properties);
            TextView tv = v.findViewById(R.id.more_info_prompt);

            if (lv.getVisibility() == View.VISIBLE) {
                lv.setVisibility(View.GONE);
                tv.setVisibility(View.VISIBLE);
                lv.setAdapter(null);
                return;
            }

            tv.setVisibility(View.GONE);
            lv.setVisibility(View.INVISIBLE);

            ListAdapter a = new ListItemAdapter(getActivity(), infos);
            lv.setAdapter(a);

            // Calculate correct height
            int totalHeight = 0;
            for (int i=0; i < a.getCount(); i++) {
                View li = a.getView(i, null, lv);
                li.measure(0, 0);
                totalHeight += li.getMeasuredHeight();
            }

            // Set correct height
            ViewGroup.LayoutParams par = lv.getLayoutParams();
            par.height = totalHeight + (lv.getDividerHeight() * (a.getCount() - 1));
            lv.setLayoutParams(par);
            lv.setVisibility(View.VISIBLE);
            lv.requestLayout();

            lv.setVisibility(View.VISIBLE);
        }
    }
}

File: src/main/java/au/id/micolous/metrodroid/fragment/CardBalanceFragment.kt
/*
 * CardBalanceFragment.kt
 *
 * Copyright 2012-2014 Eric Butler <eric@codebutler.com>
 * Copyright 2015 Michael Farrell <micolous+git@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.fragment

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.content.res.AppCompatResources
import androidx.fragment.app.ListFragment
import au.id.micolous.farebot.R
import au.id.micolous.metrodroid.activity.CardInfoActivity
import au.id.micolous.metrodroid.multi.Localizer
import au.id.micolous.metrodroid.transit.Subscription
import au.id.micolous.metrodroid.transit.TransitBalance
import au.id.micolous.metrodroid.transit.TransitData
import au.id.micolous.metrodroid.ui.ListItem

class CardBalanceFragment : ListFragment() {
    private var mTransitData: TransitData? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mTransitData = arguments!!.getParcelable(CardInfoActivity.EXTRA_TRANSIT_DATA)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val combined = ArrayList<Any>()
        val balances = mTransitData!!.balances
        if (balances != null)
            combined.addAll(balances)
        val subscriptions = mTransitData!!.subscriptions
        if (subscriptions != null)
            combined.addAll(subscriptions)
        listAdapter = BalancesAdapter(activity!!, combined)
    }

    private inner class BalancesAdapter internal constructor(context: Context, balances: List<Any>) : ArrayAdapter<Any>(context, 0, balances) {

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val item = getItem(position)

            if (item == null) {
                // https://github.com/micolous/metrodroid/issues/28
                Log.w(TAG, "null balance received -- this is an error")
                return getErrorView(convertView, parent, "null")
            }

            return (item as? TransitBalance)?.let { getBalanceView(convertView, parent, it) }
                    ?: ((item as? Subscription)?.let { getSubscriptionView(convertView, parent, it) }
                            ?: getErrorView(convertView, parent, item.javaClass.simpleName))

        }

        private fun getErrorView(convertView: View?, parent: ViewGroup, err: String): View {
            var view = convertView
            if (view == null || view.tag !== TAG_ERROR_VIEW) {
                view = activity?.layoutInflater?.inflate(R.layout.balance_item, parent, false)
                view!!.tag = TAG_ERROR_VIEW
            }

            (view.findViewById<View>(R.id.balance) as TextView).text = err
            return view
        }

        fun getSubscriptionView(convertView: View?, parent: ViewGroup, subscription: Subscription): View {
            var view = convertView
            if (view == null || view.tag !== TAG_SUBSCRIPTION_VIEW) {
                view = activity?.layoutInflater?.inflate(R.layout.subscription_item, parent, false)
                view!!.tag = TAG_SUBSCRIPTION_VIEW
            }

            val validView = view.findViewById<TextView>(R.id.valid)
            val validity = subscription.formatValidity()
            if (validity != null) {
                validView.text = validity.spanned
                validView.visibility = View.VISIBLE
            } else
                validView.visibility = View.GONE

            val tripsView = view.findViewById<TextView>(R.id.trips)
            val daysView = view.findViewById<TextView>(R.id.days)
            val remainingTrips = subscription.formatRemainingTrips()

            if (remainingTrips != null) {
                tripsView.text = remainingTrips
                tripsView.visibility = View.VISIBLE
            } else {
                tripsView.visibility = View.GONE
            }

            val remainingDays = subscription.remainingDayCount
            if (remainingDays != null) {
                daysView.text = Localizer.localizePlural(R.plurals.remaining_day_count,
                        remainingDays, remainingDays)
                daysView.visibility = View.VISIBLE
            } else {
                daysView.visibility = View.GONE
            }

            val companyView = view.findViewById<TextView>(R.id.company)
            val agencyName = subscription.getAgencyName(true)
            if (agencyName != null) {
                companyView.text = agencyName.spanned
                companyView.visibility = View.VISIBLE
            } else {
                companyView.text = ""
                companyView.visibility = View.GONE
            }
            val nameView = view.findViewById<TextView>(R.id.name)
            val name = subscription.subscriptionName
            if (name != null) {
                nameView.text = name
                nameView.visibility = View.VISIBLE
            } else {
                nameView.visibility = View.GONE
            }

            // TODO: Replace this with structured data.
            val usedView = view.findViewById<TextView>(R.id.used)
            if (subscription.subscriptionState === Subscription.SubscriptionState.UNKNOWN) {
                usedView.visibility = View.GONE
            } else {
                usedView.setText(subscription.subscriptionState.descriptionRes)
                usedView.visibility = View.VISIBLE
            }

            val paxLayout = view.findViewById<LinearLayout>(R.id.pax_layout)
            val paxIcon = view.findViewById<ImageView>(R.id.pax_icon)
            val paxTextView = view.findViewById<TextView>(R.id.pax_text_view)
            val pax = subscription.passengerCount

            if (pax >= 1) {
                paxTextView.text = "$pax"
                paxIcon.contentDescription = Localizer.localizePlural(R.plurals.passengers, pax)

                paxIcon.setImageDrawable(AppCompatResources.getDrawable(context,
                        if (pax == 1) R.drawable.material_ic_person_24dp else R.drawable.material_ic_group_24dp))

                paxLayout.visibility = View.VISIBLE
                // company and pax have the same height dictated by company
                // hence we need to show company if we want to show pax
                // even if company is empty
                companyView.visibility = View.VISIBLE
            } else {
                // No information.
                paxLayout.visibility = View.GONE
            }

            val properties = view.findViewById<ListView>(R.id.properties)
            val moreInfoPrompt = view.findViewById<TextView>(R.id.more_info_prompt)

            if (subHasExtraInfo(subscription)) {
                moreInfoPrompt.visibility = View.VISIBLE
                properties.visibility = View.GONE
            } else {
                properties.visibility = View.GONE
                moreInfoPrompt.visibility = View.GONE
            }

            return view
        }

        private fun getBalanceView(convertView: View?,
                                   parent: ViewGroup, balance: TransitBalance): View {
            var view = convertView
            if (view == null || view.tag !== TAG_BALANCE_VIEW) {
                view = activity?.layoutInflater?.inflate(R.layout.balance_item, parent, false)
                view!!.tag = TAG_BALANCE_VIEW
            }

            val validView = view.findViewById<TextView>(R.id.valid)
            val validity = TransitBalance.formatValidity(balance)
            if (validity != null) {
                validView.text = validity.spanned
                validView.visibility = View.VISIBLE
            } else
                validView.visibility = View.GONE

            val name = balance.name
            val nameView = view.findViewById<TextView>(R.id.name)
            val balanceView = view.findViewById<TextView>(R.id.balance)
            val balanceCur = balance.balance
            if (name != null) {
                nameView.text = name
                nameView.visibility = View.VISIBLE
            } else
                nameView.visibility = View.GONE

            balanceView.text = balanceCur.maybeObfuscateBalance().formatCurrencyString(true).spanned
            balanceView.visibility = View.VISIBLE

            return view
        }

        override fun isEnabled(position: Int): Boolean {
            val item = getItem(position) ?: return false

            // We don't do anything for balances, yet.
            return (item as? Subscription)?.let { subHasExtraInfo(it) } ?: false
        }
    }

    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {

        Log.d(TAG, "Clicked $id $position")
        val item = listAdapter.getItem(position) ?: return

        if (item is TransitBalance) {
            return
        }

        if (item is Subscription) {
            val infos = subMergeInfos(item) ?: return

            val lv = v.findViewById<ListView>(R.id.properties)
            val tv = v.findViewById<TextView>(R.id.more_info_prompt)

            if (lv.visibility == View.VISIBLE) {
                lv.visibility = View.GONE
                tv.visibility = View.VISIBLE
                lv.adapter = null
                return
            }

            tv.visibility = View.GONE
            lv.visibility = View.INVISIBLE

            val a = ListItemAdapter(activity!!, infos)
            lv.adapter = a

            // Calculate correct height
            var totalHeight = 0
            for (i in 0 until a.count) {
                val li = a.getView(i, null, lv)
                li.measure(0, 0)
                totalHeight += li.measuredHeight
            }

            // Set correct height
            val par = lv.layoutParams
            par.height = totalHeight + lv.dividerHeight * (a.count - 1)
            lv.layoutParams = par
            lv.visibility = View.VISIBLE
            lv.requestLayout()

            lv.visibility = View.VISIBLE
        }
    }

    companion object {
        private const val TAG = "CardBalanceFragment"

        private const val TAG_BALANCE_VIEW = "balanceView"
        private const val TAG_SUBSCRIPTION_VIEW = "subscriptionView"
        private const val TAG_ERROR_VIEW = "errorView"

        internal fun subHasExtraInfo(sub: Subscription): Boolean = Subscription.hasInfo(sub)

        internal fun subMergeInfos(sub: Subscription): List<ListItem>? = Subscription.mergeInfo(sub)
    }
}
